File: packages/sdk/test/pda-vectors.ts
/**
 * PDA vectors: derive every account address from the SDK's seed builders and
 * record the result for programs/tenet/src/pda_vectors.rs to reproduce with the
 * runtime's own `find_program_address`.
 *
 * The derivation below is test-only and Node-only (node:crypto). It exists so
 * the check needs no Solana library while D-05 is open. It is itself verified
 * by the Rust replay: a wrong on-curve test would change the bumps.
 */
import { createHash } from "node:crypto";
import { PROGRAM_ID, seeds, type Key } from "../src/pda.ts";

// ---------------------------------------------------------------- base58

const ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

export function base58Decode(s: string): Uint8Array {
  let n = 0n;
  for (const c of s) {
    const i = ALPHABET.indexOf(c);
    if (i < 0) throw new Error(`invalid base58 character: ${c}`);
    n = n * 58n + BigInt(i);
  }
  const bytes: number[] = [];
  while (n > 0n) { bytes.unshift(Number(n & 0xffn)); n >>= 8n; }
  for (const c of s) { if (c !== "1") break; bytes.unshift(0); }
  return Uint8Array.from(bytes);
}

// ---------------------------------------------------------------- ed25519

const P = (1n << 255n) - 19n;
const mod = (a: bigint) => ((a % P) + P) % P;
function pow(b: bigint, e: bigint): bigint {
  let r = 1n; b = mod(b);
  while (e > 0n) { if (e & 1n) r = mod(r * b); b = mod(b * b); e >>= 1n; }
  return r;
}
const D = mod(-121665n * pow(121666n, P - 2n));

/**
 * Whether 32 bytes decompress to an ed25519 point — the same test as
 * curve25519-dalek's `CompressedEdwardsY::decompress().is_some()`, which is what
 * Solana uses to reject a PDA candidate. The top bit is the sign of x and is
 * ignored for validity; y is reduced mod p (dalek does not reject
 * non-canonical y). Valid iff (y² − 1) / (d·y² + 1) is a square mod p.
 */
export function isOnCurve(bytes: Uint8Array): boolean {
  let y = 0n;
  for (let i = 31; i >= 0; i--) y = (y << 8n) | BigInt(bytes[i]);
  y &= (1n << 255n) - 1n;
  const y2 = mod(y * y);
  const u = mod(y2 - 1n);
  const v = mod(D * y2 + 1n);
  const w = mod(u * pow(v, P - 2n));
  return w === 0n || pow(w, (P - 1n) / 2n) === 1n;
}

export function findProgramAddress(seedList: Uint8Array[], programId: Uint8Array): [Uint8Array, number] {
  for (let bump = 255; bump >= 0; bump--) {
    const h = createHash("sha256");
    for (const s of seedList) h.update(s);
    h.update(Uint8Array.of(bump));
    h.update(programId);
    h.update("ProgramDerivedAddress");
    const addr = new Uint8Array(h.digest());
    if (!isOnCurve(addr)) return [addr, bump];
  }
  throw new Error("no viable bump");
}

// ---------------------------------------------------------------- vectors

const hex = (b: Uint8Array) => Buffer.from(b).toString("hex");

export interface PdaVector {
  fn: string;
  args: string[]; // 64-char hex for keys, decimal for u64
  address: string; // hex
  bump: number;
}

export function generatePdaVectors(perFunction = 6): { programId: string; vectors: PdaVector[] } {
  const programId = base58Decode(PROGRAM_ID);
  if (programId.length !== 32) throw new Error("PROGRAM_ID is not 32 bytes");

  let counter = 0;
  const k = (): Key => new Uint8Array(createHash("sha256").update(`tenet-pda-${counter++}`).digest());
  const ints = [0n, 1n, 255n, 256n, (1n << 32n) + 7n, (1n << 64n) - 1n];

  const vectors: PdaVector[] = [];
  const add = (fn: string, args: (Key | bigint)[], s: Uint8Array[]) => {
    const [addr, bump] = findProgramAddress(s, programId);
    vectors.push({
      fn,
      args: args.map((a) => (typeof a === "bigint" ? a.toString() : hex(a))),
      address: hex(addr),
      bump,
    });
  };

  for (let i = 0; i < perFunction; i++) {
    const [a, b] = [k(), k()];
    const n = ints[i % ints.length];
    if (i === 0) add("config", [], seeds.config()); // takes no inputs: one vector
    if (i === 0) add("devnet_test_market", [], seeds.devnetTestMarket());
    if (i === 0) add("devnet_test_mint", [], seeds.devnetTestMint());
    add("mandate", [a], seeds.mandate(a));
    add("mandate_asset", [a, b], seeds.mandateAsset(a, b));
    add("registry", [a], seeds.registry(a));
    add("circle", [a], seeds.circle(a));
    add("circle_asset", [a, b], seeds.circleAsset(a, b));
    add("vault_authority", [a], seeds.vaultAuthority(a));
    add("asset_vault", [a, b], seeds.assetVault(a, b));
    add("usdc_vault", [a], seeds.usdcVault(a));
    add("epoch", [a, n], seeds.epoch(a, n));
    add("epoch_escrow", [a, n], seeds.epochEscrow(a, n));
    add("receipt", [a, b], seeds.receipt(a, b));
    add("member", [a, b], seeds.member(a, b));
    add("nav_snapshot", [a], seeds.navSnapshot(a));
    add("redemption", [a, b, n], seeds.redemption(a, b, n));
    add("redemption_asset", [a, b], seeds.redemptionAsset(a, b));
    add("exec_auth", [a, b, n], seeds.execAuth(a, b, n));
    add("amendment", [a, n], seeds.amendment(a, n));
    add("amendment_vote", [a, b], seeds.amendmentVote(a, b));
  }
  return { programId: PROGRAM_ID, vectors };
}

export function serializePda(v: ReturnType<typeof generatePdaVectors>): string {
  return JSON.stringify({
    _comment:
      "GENERATED by scripts/gen-pda-vectors.ts from packages/sdk/src/pda.ts. Do not edit. " +
      "Replayed by programs/tenet/src/pda_vectors.rs.",
    programId: v.programId,
    count: v.vectors.length,
    vectors: v.vectors,
  }, null, 1) + "\n";
}

File: programs/tenet/src/pda_vectors.rs
//! Replay `tests/vectors/pda.json` — addresses derived from the SDK's seed
//! builders (`packages/sdk/src/pda.ts`) — through `pda.rs`, which uses the
//! runtime's own `find_program_address`.
//!
//! A mismatch means a client would address the wrong account with no error at
//! the call site. It also cross-checks the TS test harness's hand-written
//! on-curve test: a wrong curve check would show up here as a different bump.

use anchor_lang::prelude::Pubkey;
use serde_json::Value;

use crate::pda;

const VECTORS: &str = include_str!("../../../tests/vectors/pda.json");

fn key(args: &[Value], i: usize) -> Pubkey {
    let h = args[i].as_str().unwrap();
    let mut b = [0u8; 32];
    for (j, byte) in b.iter_mut().enumerate() {
        *byte = u8::from_str_radix(&h[2 * j..2 * j + 2], 16).unwrap();
    }
    Pubkey::new_from_array(b)
}

fn int(args: &[Value], i: usize) -> u64 {
    args[i].as_str().unwrap().parse().unwrap()
}

fn hex(p: &Pubkey) -> String {
    p.to_bytes().iter().map(|b| format!("{b:02x}")).collect()
}

#[test]
fn vectors_were_generated_for_this_program_id() {
    // If the program keypair is ever regenerated, every PDA moves. The vectors
    // must be regenerated with it, not left describing a different program.
    let doc: Value = serde_json::from_str(VECTORS).unwrap();
    assert_eq!(doc["programId"].as_str().unwrap(), crate::ID.to_string());
}

#[test]
fn rust_derivations_match_the_sdk_seed_builders() {
    let doc: Value = serde_json::from_str(VECTORS).expect("tests/vectors/pda.json parses");
    let vectors = doc["vectors"].as_array().unwrap();
    assert_eq!(vectors.len() as u64, doc["count"].as_u64().unwrap());

    let mut failures = Vec::new();
    for (i, v) in vectors.iter().enumerate() {
        let f = v["fn"].as_str().unwrap();
        let a = v["args"].as_array().unwrap();
        let (addr, bump) = match f {
            "config" => pda::config(),
            "mandate" => pda::mandate(&key(a, 0)),
            "mandate_asset" => pda::mandate_asset(&key(a, 0), &key(a, 1)),
            "registry" => pda::registry(&key(a, 0)),
            "circle" => pda::circle(&key(a, 0)),
            "circle_asset" => pda::circle_asset(&key(a, 0), &key(a, 1)),
            "vault_authority" => pda::vault_authority(&key(a, 0)),
            "asset_vault" => pda::asset_vault(&key(a, 0), &key(a, 1)),
            "usdc_vault" => pda::usdc_vault(&key(a, 0)),
            "epoch" => pda::epoch(&key(a, 0), int(a, 1)),
            "epoch_escrow" => pda::epoch_escrow(&key(a, 0), int(a, 1)),
            "receipt" => pda::receipt(&key(a, 0), &key(a, 1)),
            "member" => pda::member(&key(a, 0), &key(a, 1)),
            "nav_snapshot" => pda::nav_snapshot(&key(a, 0)),
            "redemption" => pda::redemption(&key(a, 0), &key(a, 1), int(a, 2)),
            "redemption_asset" => pda::redemption_asset(&key(a, 0), &key(a, 1)),
            "exec_auth" => pda::exec_auth(&key(a, 0), &key(a, 1), int(a, 2)),
            "amendment" => pda::amendment(&key(a, 0), int(a, 1)),
            "amendment_vote" => pda::amendment_vote(&key(a, 0), &key(a, 1)),
            other => panic!("vector {i}: unknown derivation {other}"),
        };
        let want_addr = v["address"].as_str().unwrap();
        let want_bump = v["bump"].as_u64().unwrap() as u8;
        if hex(&addr) != want_addr || bump != want_bump {
            failures.push(format!(
                "#{i} {f}: rust {}/{bump}, sdk {want_addr}/{want_bump}",
                hex(&addr)
            ));
        }
    }
    assert!(
        failures.is_empty(),
        "{} mismatches:\n{}",
        failures.len(),
        failures.join("\n")
    );

    // Guard against a vacuous pass: some vectors must need a bump below 255,
    // i.e. the first candidate was ON the curve and had to be rejected.
    assert!(
        vectors.iter().any(|v| v["bump"].as_u64().unwrap() < 255),
        "no vector exercised on-curve rejection"
    );
}
